### ISSUE TITLE
Problem with batch execution

### ISSUE THREAD

**Issue by msciciel** [saltstack/salt#18358]
**salt '*' test.ping** : is executed on most of hosts, to be precise on 750 from 1150 minions (it's another issue but is not subject of this issue)
**salt -b 20 '*' test.ping** : is matched to be executed only on 20-100 minions

If execute with more precise matching for example:
**salt -b 20 'xen*' test.ping** then all hosts are corrected matched for batch execution.

My salt version is:

```
           Salt: 2014.7.0
         Python: 2.6.6 (r266:84292, Jan 22 2014, 09:42:36)
         Jinja2: unknown
       M2Crypto: 0.20.2
 msgpack-python: 0.1.13
   msgpack-pure: Not Installed
       pycrypto: 2.0.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.3.1
           RAET: Not Installed
            ZMQ: 3.2.4
           Mako: Not Installed
```

My salt-master configuration is:

```
gitfs_provider: GitPython
max_open_files: 8192
worker_threads: 64
timeout: 30
loop_interval: 180
job_cache: True
minion_data_cache: True
open_mode: True
auto_accept: True
state_verbose: false
fileserver_backend:
   - git
gitfs_remotes:
ext_pillar:
log_level_logfile: warning
reactor:
  - 'minion_start':
    - /srv/reactor/sync_grains.sls
```

I think that there wasn't such problem in 2014.1.X release.

My salt-master machine is:

```
 16 cores of Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz
 16GB RAM
 CentOS release 6.6 (Final) 64bit
```

**Comment by cachedout** [comment]
Glad to hear it. :]

**Comment by msciciel** [comment]
When It would be included in salt release ? After upgrade to 2014.7.1 i had to apply patch manually again :(

**Comment by cachedout** [comment]
That fix should be in 2015.2 which currently has a release candidate out.

**Comment by rallytime** [comment]
@msciciel It actually looks like this bug was fixed on the develop branch. I will see if I can get this backported to an older branch today and let you know.

**Comment by rallytime** [comment]
@msciciel Ok, I created #21023 to backport the fix to the 2014.7 branch. This fix should be available in the next bug fix on the 2014.7 branch (2014.7.3). It will also be available in 2015.2.0 as @cachedout mentioned above.

**Comment by msciciel** [comment]
Great news. Thanks.

**Comment by jacksontj** [comment]
@msciciel The change @cachedout made uses CKMinions, which actually doesn't do what you want in this scenario. CKMinions is intended to give a super-set of minions-- and in quite a few cases returns all minions. I Made a new PR (#22350) which reverts most of this-- but should be better than the old implementation without breaking all of batching ;)

**Comment by msciciel** [comment]
New implementation is very slow if i want to execute something on 1500 minion then i need to wait a lot. Previous implementation was starting jobs almost instantly.

**Comment by jacksontj** [comment]
Could you provide logs or timing? We can't go back to the old way since the
old way just always batched across all hosts regardless of target (for most
targets).

**Comment by msciciel** [comment]
I'm executing:
time salt '*' -b 100 test.ping

And execution starts after:
old way: 0m2.335s
new way: 0m41.688s

I don't understand why minions list could not be fetched just from keys list and all minions must be pinged to be added to batch list.

Old way works great. I didn't noticed that batch is executed on all hosts regardless the defined target. I select minions very ofter with pillar oraz grains and it's always works as i expected.

**Comment by jacksontj** [comment]
That is definitely too slow... What is your "timeout" set to on the master?

**Comment by msciciel** [comment]
Timeout is 30 seconds. But still i dont understand why list of minions must be fetched by ping and not just from keys list without minion ping.

**Comment by msciciel** [comment]
Any chances for quicker batch job start (instant start was better than 10-30s delay) ? In addition i see that current implementation in version 2015.5.2 starts first batch in parallel, then next jobs in single mode and after few hosts crash occurs:

<pre>
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
ValueError: list.remove(x): x not in list
Traceback (most recent call last):
  File "/usr/bin/salt", line 10, in <module>
    salt_main()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 349, in salt_main
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/salt.py", line 98, in run
    for res in batch.run():
  File "/usr/lib/python2.6/site-packages/salt/cli/batch.py", line 179, in run
    active.remove(minion)
ValueError: list.remove(x): x not in list
Traceback (most recent call last):
  File "/usr/bin/salt", line 10, in <module>
    salt_main()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 349, in salt_main
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/salt.py", line 98, in run
    for res in batch.run():
  File "/usr/lib/python2.6/site-packages/salt/cli/batch.py", line 179, in run
    active.remove(minion)
ValueError: list.remove(x): x not in list
</pre>

**Comment by marnovdm** [comment]
I'm having the same issue as msciciel (list.remove(x): x not in list) when batching in latest 2015.5.2 release on Debian 8.

**Comment by msciciel** [comment]
Is there chances for working batching without update to 2015.8 ? Version 2015.5.5 is useless in that matter.
